source/common/router/router.cc

```diff
@@ -662,7 +664,9 @@ void Filter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callb
   // As the decoder filter only pushes back via watermarks once data has reached
   // it, it can latch the current buffer limit and does not need to update the
   // limit if another filter increases it.
-  buffer_limit_ = callbacks_->decoderBufferLimit();
+  if (callbacks_->decoderBufferLimit() != 0) {
+    retry_shadow_buffer_limit_ = callbacks_->decoderBufferLimit();
```

Comment:
its a little bit hard to follow that this limit is set to the buffer limit in the callbacks, then potentially updated in `decodeHeaders` to account for the retry/shadow limit. Maybe a comment here would help?